topdown/eval.go

```diff
@@ -1542,15 +1551,37 @@ func (e evalFunc) eval(iter unifyIterator) error {
 	return nil
 }
 
-func (e evalFunc) evalOneRule(iter unifyIterator, rule *ast.Rule, prev *ast.Term) (*ast.Term, error) {
+func (e evalFunc) evalCache(argCount int, iter unifyIterator) (ast.Ref, bool, error) {
+	var plen int
+	if len(e.terms) == argCount+2 { // func name + output = 2
+		plen = len(e.terms) - 1
+	} else {
+		plen = len(e.terms)
+	}
+	cacheKey := make([]*ast.Term, plen)
+	for i := 0; i < plen; i++ {
+		cacheKey[i] = e.e.bindings.Plug(e.terms[i])
+	}
+
+	cached := e.e.virtualCache.Get(cacheKey)
+	if cached != nil {
+		e.e.instr.counterIncr(evalOpVirtualCacheHit)
+		if argCount == len(e.terms)-1 && ast.Boolean(false).Equal(cached.Value) {
+			return nil, true, nil
+		}
+		// TODO: update bindings?
```

Comment:
There are two cases to deal with here:

1. The call-site does not capture the output value (e.g., the call is `f(1)` and the definition is `f(x) = y { ... }`). In this case, we need to test if the result is false (which we're currently doing on L1569). If the result is false, then return nil and DO NOT invoke the iterator. If the result is NOT false then invoke `iter` to continue evaluation.

2. The call-site does capture the output value (e.g., the call is `f(1, x)` and the definition is `f(x) = y { ... }`). In this case, we need to unify the output value (`x`) with the result. In the normal code path this is handled by the `child.biunifyArrays(...)` call.

In pseudocode:

```
// if output value is NOT captured...
if argCount == len(e.terms)-1 {
   if cached value is false {
      return nil
   }
   return iter()
}

// otherwise output value is captured...
return e.e.unify(e.terms[len(e.terms)-1], cached, iter)
```